flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsPartitioning.java

```diff
@@ -0,0 +1,109 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.connector.sink.abilities;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+
+import java.util.Map;
+
+/**
+ * Enables to write partitioned data in a {@link DynamicTableSink}.
+ *
+ * <p>Partitions split the data stored in an external system into smaller portions that are identified
+ * by one or more string-based partition keys. A single partition is represented as a {@code Map < String, String >}
+ * which maps each partition key to a partition value. Partition keys and their order are defined by the
+ * catalog table.
+ *
+ * <p>For example, data can be partitioned by region and within a region partitioned by month. The order
+ * of the partition keys (in the example: first by region then by month) is defined by the catalog table. A
+ * list of partitions could be:
+ * <pre>
+ *   List(
+ *     ['region'='europe', 'month'='2020-01'],
+ *     ['region'='europe', 'month'='2020-02'],
+ *     ['region'='asia', 'month'='2020-01'],
+ *     ['region'='asia', 'month'='2020-02']
+ *   )
+ * </pre>
+ *
+ * <p>Given the following partitioned table:
+ * <pre>{@code
+ *   CREATE TABLE t (a INT, b STRING, c DOUBLE, region STRING, month STRING) PARTITION BY (region, month);
+ * }</pre>
+ *
+ * <p>We can insert data into <i>static table partitions</i> using the {@code INSERT INTO ... PARTITION} syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe', month='2020-01') SELECT a, b, c FROM my_view;
+ * }</pre>
+ *
+ * <p>If all partition keys get a value assigned in the {@code PARTITION} clause, the operation is considered
+ * as an "insertion into a static partition". In the above example, the query result should be written
+ * into the static partition {@code region='europe', month='2020-01'} which will be passed by the planner
+ * into {@link #applyStaticPartition(Map)}.
+ *
+ * <p>Alternatively, we can insert data into <i>dynamic table partitions</i> using the SQL syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe') SELECT a, b, c, month FROM another_view;
+ * }</pre>
+ *
+ * <p>If only a subset of all partition keys (a prefix part) get a value assigned in the {@code PARTITION}
```

Comment:
What if there is no `INTERT INTO t **PARTITION**`? If I am not mistaken the dynamic partitioning still should be applied.

Reply:
Good point. From my understand, this syntax is used to keep align with Hive partition usage which makes the the partition fields more explicit. cc @JingsongLi

Reply:
If all partition columns take dynamic values, the `PARTITION` specification will be omitted. This is different from Hive syntax which requires all partition columns in the `PARTITION` spec, both static and dynamic ones.

Reply:
In Hive 3.1, https://issues.apache.org/jira/browse/HIVE-18792 partition clause is optional too. This is a more standard way.